src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -191,109 +242,222 @@ public static int GetMaxDecodedFromUtf8Length(int length)
+                    destIndex += 1;
+                }
+
+            DoneExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.Done;
+
+            InvalidExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.InvalidData;
             }
-            else if (i2 != EncodingPad)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
+        {
+            // The JIT won't hoist these "constants", so help it
+            Vector256<sbyte> lutHi = ReadVector<Vector256<sbyte>>(s_avxDecodeLutHi);
+            Vector256<sbyte> lutLo = ReadVector<Vector256<sbyte>>(s_avxDecodeLutLo);
+            Vector256<sbyte> lutShift = ReadVector<Vector256<sbyte>>(s_avxDecodeLutShift);
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = ReadVector<Vector256<sbyte>>(s_avxDecodeShuffleVec);
+            Vector256<int> permuteVec = ReadVector<Vector256<sbyte>>(s_avxDecodePermuteVec).AsInt32();
+
+            byte* src = srcBytes;
+            byte* dest = destBytes;
+
+            //while (remaining >= 45)
+            do
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                AssertRead<Vector256<sbyte>>(src, srcStart, sourceLength);
+                Vector256<sbyte> str = Avx.LoadVector256(src).AsSByte();
 
-                i2 <<= 6;
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLo, loNibbles);
+                Vector256<sbyte> zero = Vector256<sbyte>.Zero;
 
-                i0 |= i2;
+                // https://github.com/dotnet/coreclr/issues/21247
+                if (Avx2.MoveMask(Avx2.CompareGreaterThan(Avx2.And(lo, hi), zero)) != 0)
+                    break;
 
-                if (i0 < 0)
-                    goto InvalidExit;
-                Unsafe.Add(ref bufferBytes, destIndex) = (byte)(i0 >> 16);
-                Unsafe.Add(ref bufferBytes, destIndex + 1) = (byte)(i0 >> 8);
-                destIndex += 2;
+                Vector256<sbyte> eq2F = Avx2.CompareEqual(str, mask2F);
+                Vector256<sbyte> shift = Avx2.Shuffle(lutShift, Avx2.Add(eq2F, hiNibbles));
+                str = Avx2.Add(str, shift);
+
+                Vector256<short> merge_ab_and_bc = Avx2.MultiplyAddAdjacent(str.AsByte(), shuffleConstant0);
```

Comment:
A more descriptive name than `shuffleConstant0` would be nice. Possibly something that explains how the operands are being shuffled.